tools/outline_proxy_controller/network_monitor.cpp

```diff
@@ -0,0 +1,139 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <array>
+#include <cstddef>      // size_t, byte
+#include <cstring>      // memset
+#include <system_error>
+#include <type_traits>  // underlying_type_t, is_integral
+
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+
+#include <boost/asio.hpp>
+
+#include "logger.h"
+#include "network_monitor.h"
+
+using namespace outline;
+
+static const std::size_t kRecvBufferSize = 4096;
+
+namespace outline {
+
+static_assert(std::is_integral<std::underlying_type_t<NetworkChangeEvent>>::value);
+
+NetworkChangeEvent operator|(NetworkChangeEvent lhs, NetworkChangeEvent rhs) noexcept {
+  using IntType = std::underlying_type_t<NetworkChangeEvent>;
+  return static_cast<NetworkChangeEvent>(static_cast<IntType>(lhs) | static_cast<IntType>(rhs));
+}
+
+NetworkChangeEvent& operator|=(NetworkChangeEvent& lhs, NetworkChangeEvent rhs) noexcept {
+  return lhs = lhs | rhs;
+}
+
+NetworkChangeEvent operator&(NetworkChangeEvent lhs, NetworkChangeEvent rhs) noexcept {
+  using IntType = std::underlying_type_t<NetworkChangeEvent>;
+  return static_cast<NetworkChangeEvent>(static_cast<IntType>(lhs) & static_cast<IntType>(rhs));
+}
+
+std::string to_string(NetworkChangeEvent evt) {
+  if (evt == NetworkChangeEvent::kNone) {
+    return "None";
+  }
+  std::string result{};
+  if ((evt & NetworkChangeEvent::kNicChanged) == NetworkChangeEvent::kNicChanged) {
+    result += "NIC";
+  }
+  if ((evt & NetworkChangeEvent::kAddressChanged) == NetworkChangeEvent::kAddressChanged) {
+    result += (result.empty() ? "IP" : " IP");
+  }
+  if ((evt & NetworkChangeEvent::kRouteChanged) == NetworkChangeEvent::kRouteChanged) {
+    result += (result.empty() ? "Route" : " Route");
+  }
+  return result;
+}
+
+}
+
+NetworkMonitor::NetworkMonitor(const boost::asio::any_io_executor &io_context)
+  : netlink_socket_{io_context}
+{
+  using namespace boost::asio::generic;
+
+  sockaddr_nl sa;
+  std::memset(&sa, 0, sizeof(sa));
+  sa.nl_family = AF_NETLINK;
+  sa.nl_groups = RTMGRP_LINK         // network card change
+               | RTMGRP_IPV4_IFADDR  // IPv4 address change
+               | RTMGRP_IPV4_ROUTE   // IPv4 route table change
+               | RTMGRP_IPV6_IFADDR  // IPv6 address change
+               | RTMGRP_IPV6_ROUTE;  // IPv6 route table change
+
+  netlink_socket_.open({AF_NETLINK, NETLINK_ROUTE});
+  netlink_socket_.bind({&sa, sizeof(sa)});
+
+  logger.info("network monitor initialized");
```

Comment:
Can the initialization fail? If so, we may want to have a factory function that pre-initializes the netlink_socket and passes it to the constructor.

Also, passing a socket to the class makes it more reusable and testeable.

Reply:
Yes it might throw (otherwise I will add `noexcept` to the function signature). But this is totally expected and no resources will be leaked because I am using the [RAII idiom](https://en.cppreference.com/w/cpp/language/raii) here, and it is recommended to throw exceptions in constructor:

> encapsulate each resource into a class, where the constructor acquires the resource and establishes all class invariants or throws an exception if that cannot be done

Although Google seems to [against exceptions](https://google.github.io/styleguide/cppguide.html#Exceptions) and RAII, I have reservations about it, also [C++ creator supports RAII and exceptions](https://www.stroustrup.com/bs_faq2.html#exceptions-why).